akka-stream/src/main/scala/akka/stream/impl/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,117 @@
+  def props(f: File, chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  private final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, chunkSize: Int, initialBuffer: Int, maxBuffer: Int) extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  // TODO also have a Promise[Long] here? for bytes-read?
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
+
+  private val fs = new RandomAccessFile(f, "r") // best way to express this in JDK6, OpenOption are available since JDK7
+  private var chan: FileChannel = _
+
+  override def preStart() = {
+    try {
+      chan = fs.getChannel
+    } catch {
+      case ex: Exception ⇒
+        log.error(ex, "Unable to open channel for {}", f.getAbsolutePath)
+        context.stop(self)
+    }
+
+    readAndSignal(initialBuffer)
+
+    super.preStart()
+  }
+
+  def receive = {
+    case ActorPublisherMessage.Request(elements) ⇒ readAndSignal(maxBuffer)
+    case Continue                                ⇒ if (totalDemand > 0) readAndSignal(maxBuffer)
+    case ActorPublisherMessage.Cancel            ⇒ context.stop(self)
+  }
+
+  def readAndSignal(readAhead: Int): Unit =
+    if (isActive) {
+      // signal from available buffer right away
+      signalOnNexts()
+
+      // read chunks until readAhead is fulfilled
+      while (availableChunks.length < readAhead && !eofEncountered)
+        loadChunk()
+
+      self ! Continue
+    }
+
+  private def signalOnNexts(): Unit =
+    if (availableChunks.nonEmpty && totalDemand > 0) {
+      val ready = availableChunks.head
+      availableChunks = availableChunks.tail
+
+      onNext(ready)
+
+      if (totalDemand > 0) signalOnNexts()
+    } else if (eofEncountered) onComplete()
+
+  /** BLOCKING I/O READ */
+  def loadChunk() = {
+    val buf = buffs.acquire()
+    buf.flip().clear()
+
+    // blocking read
+    val readBytes = chan.read(buf)
+
+    readBytes match {
+      case -1 | 0 ⇒
+        // had nothing to read into this chunk, will complete now
+        eofReachedAtOffset = chan.position
+        log.debug("No more bytes available to read (got `-1` from `read`), marking final bytes of file @ " + eofReachedAtOffset)
+
+      case _ ⇒
+        buf.flip()
+        readBytesTotal += readBytes
+        availableChunks :+= ByteString(buf).take(readBytes)
+        buffs.release(buf)
+
+      // valid read, continue
```

Comment:
stopping supervision is the only that makes sense and that would require another actor, so I would say catch and stop